esmvalcore/preprocessor/_regrid_esmpy.py

```diff
@@ -39,6 +40,130 @@
 # }
 
 
+class _ESMPyRegridder:
+    """General ESMPy regridder.
+
+    Parameters
+    ----------
+    src_cube:
+        Cube defining the source grid.
+    tgt_cube:
+        Cube defining the target grid.
+    method:
+        Regridding algorithm.
+    mask_threshold:
+        Threshold used to regrid mask of input cube.
+
+    """
+
+    def __init__(
+        self,
+        src_cube: Cube,
+        tgt_cube: Cube,
+        method: str = 'linear',
+        mask_threshold: float = 0.99,
+    ):
+        """Initialize class instance."""
+        self.src_cube = src_cube
+        self.tgt_cube = tgt_cube
+        self.method = method
+        self.mask_threshold = mask_threshold
+
+    def __call__(self, cube: Cube) -> Cube:
+        """Perform regridding.
+
+        Parameters
+        ----------
+        cube:
+            Cube to be regridded.
+
+        Returns
+        -------
+        Cube
+            Regridded cube.
+
+        """
+        src_rep, dst_rep = get_grid_representants(cube, self.tgt_cube)
+        regridder = build_regridder(
+            src_rep, dst_rep, self.method, mask_threshold=self.mask_threshold
+        )
+        result = map_slices(cube, regridder, src_rep, dst_rep)
+        return result
+
+
+class _ESMPyScheme:
+    """General irregular regridding scheme.
+
+    This class can be used in :meth:`iris.cube.Cube.regrid`.
+
+    Note
+    ----
+    See `ESMPy <http://www.earthsystemmodeling.org/
+    esmf_releases/non_public/ESMF_7_0_0/esmpy_doc/html/
+    RegridMethod.html#ESMF.api.constants.RegridMethod>`__ for more details on
+    this.
+
+    Parameters
+    ----------
+    mask_threshold:
+        Threshold used to regrid mask of source cube.
+
+    """
+
+    _METHOD = ''
+
+    def __init__(self, mask_threshold: float = 0.99):
+        """Initialize class instance."""
+        self.mask_threshold = mask_threshold
+
+    def __repr__(self) -> str:
+        """Return string representation of class."""
+        return (
+            f'{self.__class__.__name__}(mask_threshold={self.mask_threshold})'
+        )
+
+    def regridder(self, src_cube: Cube, tgt_cube: Cube) -> _ESMPyRegridder:
+        """Get regridder.
+
+        Parameters
+        ----------
+        src_cube:
+            Cube defining the source grid.
+        tgt_cube:
+            Cube defining the target grid.
+
+        Returns
+        -------
+        _ESMPyRegridder
+            Regridder instance.
+
+        """
+        return _ESMPyRegridder(
+            src_cube,
+            tgt_cube,
+            method=self._METHOD,
+            mask_threshold=self.mask_threshold,
+        )
```

Comment:
OK call me old fashioned but to me this is a bit wrong: public method in a private class is something I don't agree with, especially since the `ESMPYScheme` name doesn't reflect the fact that an actual regridder instance is built within its bellows. What's wrong with the old generic and public `regrid` wrapper that the user is exposed to?

Reply:
I agree with you that this looks too complicated, but this is how iris expects regridding schemes to look like! The scheme class needs a public method `regridder(self, src_cube, target_grid)` that returns an object of a regridder class. This regridder class needs to have a method `def __call__(self, cube)` that actually performs the regridding. [Here](https://github.com/SciTools/iris/blob/8694479607d5967561affa05b85a6b8fb49477df/lib/iris/analysis/__init__.py#L3083-L3125) is an example for a regridder in iris.

The advantage of this new syntax is that it's much more generic. All of our regridding schemes are used the same way (within `iris.cube.Cube.regrid()`. In the old version, some schemes used the iris syntax, some use this `regrid` function that you mentiond, etc.

Reply:
indeed - I caught that, and liked it - but here's the thing - aligning ourselves to iris is not something we should aim for. If we find a nice, generic, and robust way of writing code and functionality we should go that way, regardless iris do other things; in this case, it's prob useful since we are in essence wrapping our regridding around iris regridding, so I'll grab the bait :fish: